src/meili/match_route.cc

```diff
@@ -141,15 +141,21 @@ void cut_segments(const std::vector<MatchResult>& match_results,
     if (!curr_match.is_break_point && curr_idx != last_idx) {
       continue;
     }
-    // we want to handle to loop by locating the correct target edge by comparing the distance alone
-    bool loop = prev_match.edgeid == curr_match.edgeid &&
-                prev_match.distance_along > curr_match.distance_along;
+    // we want to handle to loop by locating the correct target edge by comparing the distance alone.
+    // Note: there should be a proper tolerance check to see if the 'percent_along's are close
+    // to one another, but comparing percentages isn't geometrically correct. Some refactoring
+    // would be required if we want to strenghen this condition.
+    bool loop =
+        prev_match.edgeid == curr_match.edgeid && prev_match.percent_along > curr_match.percent_along;
     // if it is a loop, we start the search after the first edge
     auto last_segment = std::find_if(first_segment + static_cast<size_t>(loop), segments.end(),
                                      [&curr_match](const EdgeSegment& segment) {
                                        return (segment.edgeid == curr_match.edgeid);
                                      });
+
     assert(last_segment != segments.cend());
+    if (last_segment == segments.cend())
+      continue;
```

Comment:
the assert above is basically saying this should never happen. the continue feels like a band aide. can you explain the purpose. i havent looked at this code since we refactored it last summer so id need to spend a day coming back up to speed

Reply:
Ah yes.  A couple things going on here.  Remember the nature of the problem leads us to this condition being met.  The assert does fire - but only in a debug build.  As you say, the assert is catching a logic condition that is truly unexpected.  My fix - using ```snap_tol``` when snapping - addresses the problem.  As a result, the assert no longer occurs and you could argue that this condition is no longer needed.

However!  You know how code evolves... an upstream algorithm changes, an unexpected geometry arises... and boom, the unexpected thing starts happening again.  The assert will only "help" in a debug build.  I feel like it's a good practice to be defensive in production builds too, especially if we are avoiding a segfault.  That said, don't get me wrong, I'm not a fan of hyper-defensive code either (you know, checking for null from every call to ```new```).  This just seems like one of those places where having both an assert and a runtime check is appropriate.

Reply:
my point is that the assertion is a precondition. this `if` statement directly contradicts that. the assertion is like saying, this cannot be the case if the code up to this point is working properly. but then directly after that you say actually its cool we'll just continue.

you can think of the assertion in this case is like `std::logic_error`

Reply:
@ktatterso you're right from the one side: we definitely don't want segfaults. But, in case we continue after `assert` actually we implicitly hide the real problem. so, I'd rather remove these 2 lines or throw an exception instead of `assert`

Reply:
I agree we should throw logic error